### PR TITLE
fix(nuxt): disable View Transitions when `prefers-reduced-motion: reduce`

### LINKED ISSUE #22291: Disable View Transitions when prefers-reduced-motion is reduce
### Describe the feature

Currently the experimental view transition API doesn't honour the [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) user preference. A simple check should be added to the plugin to early return when that media query is set to reduce.

### Additional information

- [X] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

### PR DESCRIPTION
**Pull request by lukewarlow**
### 🔗 Linked issue

https://github.com/nuxt/nuxt/issues/22291

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

View transitions now early return when the `prefers-reduced-motion` user preference is reduce. This improves the out of the box accessibility of the feature.

Resolves #22291 

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/app/plugins/view-transitions.client.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { useRouter } from '#app/composables/router'
 import { defineNuxtPlugin } from '#app/nuxt'
 
 export default defineNuxtPlugin((nuxtApp) => {
-  if (!document.startViewTransition) { return }
+  if (!document.startViewTransition || window.matchMedia('(prefers-reduced-motion: reduce)').matches) { return }
```
[the remark this answers is not in the document]
Comment (lukewarlow) [reply packages/nuxt/src/app/plugins/view-transitions.client.ts:5]: I couldn't get disabling them in CSS to work myself.

Also view transitions are designed to be a pure enhancement as mentioned in the spec. So I would consider them unnecessary motion when it comes to preferring reduced motion.

Lastly unfortunately a fact of life is that most developers simply won't account for this leading to less accessible sites so it'd be nice for the framework to handle it.

**Comment by DamianGlowala** [comment]
This might help:

**Comment by lukewarlow** [comment]
> This might help:
> 
> https://developer.chrome.com/docs/web-platform/view-transitions/#reacting-to-the-reduced-motion-preference
> 
> > However, a preference for 'reduced motion' doesn't mean the user wants no motion. Instead of the above, you could choose a more subtle animation, but one that still expresses the relationship between elements, and the flow of data.

Ah thanks god knows how I missed that when I read through that post. I was missing the -group pseudo. That appears to be working now.


I still feel it's valuable to have an option to skip the transition (it skips the steps of generating the old and new images of DOM) but I get if you would rather this not be the default. I'm not familar with the plugin API does nuxtApp expose an options object?

**Comment by lukewarlow** [comment]
https://github.com/withastro/astro/pull/7757 fwiw astro appear to have gone the route of disabling view transitions by default (they used CSS instead)

File: packages/nuxt/src/app/plugins/view-transitions.client.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielroe) [reply packages/nuxt/src/app/plugins/view-transitions.client.ts:5]: We could update `experimental.viewTransition` to accept either `true | false | 'always'`. `true` will disable (as above) if `window.matchMedia('(prefers-reduced-motion: reduce)')`. But `always` will allow developers to handle the CSS for reduced motion themselves.

What do you think?

[the remark this answers is not in the document]
Comment (lukewarlow) [reply packages/nuxt/src/app/plugins/view-transitions.client.ts:5]: That sounds good I'll make that change later today.

File: packages/nuxt/src/app/plugins/view-transitions-safe.client.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { defineNuxtPlugin } from '#app/nuxt'
+import plugin from '#app/plugins/view-transitions.client'
+
+export default defineNuxtPlugin((nuxtApp) => {
+  if (!document.startViewTransition || window.matchMedia('(prefers-reduced-motion: reduce)').matches) { return }
```
Comment (lukewarlow) [inline packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: Making a separate safe plugin seemed like the best way to do this? Plugins don't seem to have any access to the nuxt config option?

File: packages/nuxt/src/app/plugins/view-transitions-safe.client.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+
+  plugin(nuxtApp)
```
Comment (lukewarlow) [inline packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:7]: Should this return the plugin?

File: packages/nuxt/src/app/plugins/view-transitions-safe.client.ts (hunk 2 above)
Comment (danielroe) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: We can specifically expose it via this template:

https://github.com/nuxt/nuxt/blob/09161d005da48b0ba82308093e509a53d9a8b899/packages/nuxt/src/core/templates.ts#L367-L378

... and then import the value from '#build/nuxt.config.mjs' as we do here:

https://github.com/nuxt/nuxt/blob/d7b11164434e5ae318cd8cf85417362fddb99d71/packages/nuxt/src/app/plugins/payload.client.ts#L7-L8

Comment (danielroe) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: (Alternatively, the approach you've taken here would work too.)

Comment (lukewarlow) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: Is there a preference for one or the other I'm happy to change it?

Comment (danielroe) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: Either way works for me 👍

Comment (lukewarlow) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: Okay in that case this should be ready to go when the team are :)

Comment (danielroe) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: actually, let me backpedal on this.

the reason is that we're doing more build-time optimisations by parsing plugins, and importing and running a plugin might cause issues in future.

if we have one plugin and read the config option from `#build/nuxt.config.mjs` that would be better

Comment (lukewarlow) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts:5]: No worries I'll get that changed!

**Comment by lukewarlow** [comment]
Apologies it's taken so long to get back to this. Hopefully the latest commit is correct

File: packages/nuxt/src/app/plugins/view-transitions-safe.client.ts (hunk 2 above)
Comment (lukewarlow) [reply packages/nuxt/src/app/plugins/view-transitions-safe.client.ts]: I think this is done correctly now :)

**Review by danielroe — APPROVED** [review]

**Review by TheAlexLichter — APPROVED** [review]

**Comment by lukewarlow** [comment]
Just double checking is this waiting on me for anything? No worries if it just needs time or other reviews, wanna make sure i'm not the blocker though.

**Comment by danielroe** [comment]
Nope! It's good to go, and I plan to merge it in time for the next minor release.

**Pull request merged**